sklearn/ensemble/_stacking.py

```diff
@@ -475,7 +484,7 @@ def fit(self, X, y, sample_weight=None):
         self.classes_ = self._le.classes_
         return super().fit(X, self._le.transform(y), sample_weight)
 
-    @if_delegate_has_method(delegate="final_estimator_")
+    @available_if(_estimator_has("predict"))
     def predict(self, X, **predict_params):
```

Comment:
Along the line of another thread, I wonder what should be the expected behavior of the following code snippet? It constructs a `StackingClassifier` with `final_estimator=None` and immediately calls `.predict` without fitting. Currently (master/main as well as in this PR), the code raises `AttributeError` instead of `NotFittedError` because of the delayed calling of `_validate_final_estimator`. Is it expected?
```
from sklearn.datasets import load_iris
from sklearn.ensemble import RandomForestClassifier, StackingClassifier
from sklearn.svm import LinearSVC
X, y = load_iris(return_X_y=True)
estimators = [
    ('rf', RandomForestClassifier(n_estimators=10, random_state=42)),
    ('svr', LinearSVC(random_state=42))
]
stacked_classifier = StackingClassifier(estimators, final_estimator=None)

stacked_classifier.predict(X)    # AttributeError instead of NotFittedError
```

Reply:
@jnothman I guess by extending the previous reasoning, this unfitted stacked classifier should be able to figure out that it supports the `predict` function as well. A quick fix could be to call `_validate_final_estimator` once in `__init__`.

Reply:
We cannot call any validation in `__init__` unfortunately (it will break things in `GridSearchCV`).

One workaround would be to check first if `estimators_` exist. If not, the `StackingClassifier` is for sure unfitted.

Reply:
Right, right. It would also break consistency with `check_is_fitted`.

I guess if we want the above unfitted stacked classifier (constructed with `final_estimator=None`) to know that it supports the `predict` function, then we need to somehow let `_estimator_has` know what the default replacement is for `final_estimator`, which in this case we know should be `LogisticRegression()`. Currently, `_validate_final_estimator` does 3 things - (1) resolves the default final estimator; (2) clones it; and (3) sets `final_estimator_`. So I think maybe we could just move step (1) to `__init__`, something similar to the following? 
```
def replace_if_none(obj, default):
    return default if obj is None else obj

class StackingClassifier(...):
  def __init__(self, estimators, final_estimator=None *, ...):
        super().__init__(
            estimators=estimators,
            final_estimator=replace_if_none(final_estimator, LogisticRegression()),    # <--
            cv=cv,
            stack_method=stack_method,
            n_jobs=n_jobs,
            passthrough=passthrough,
            verbose=verbose,
        )
```

Or, if this leads to undesirable inconsistency between input arg `final_estimator` and attr `self.final_estimator`, then maybe we can just have a `self._default_final_estimator=LogisticRegression()` in `__init__`?

Reply:
The contract in scikit-learn is whatever parameter passed in `__init__` is then affected to a class attribute with the same name: https://scikit-learn.org/dev/developers/develop.html#parameters-and-init

We should really not do any validation there.

Reply:
I see. What about the second option - just setting `self._default_final_estimator=LogisticRegression()` in `__init__`, an private attr that the validation function later can use?